### ISSUE TITLE
Support signing payloads with custom signedExtensions

### ISSUE THREAD

**Issue by KarishmaBothara** [polkadot-js/extension#631]
Support to sign transaction with userExtensions (signedExtension specific/custom to chains) .
Right now it injects the current signed extensions for encoding - https://github.com/polkadot-js/extension/blob/master/packages/extension-base/src/background/RequestExtrinsicSign.ts#L19

    ` registry.setSignedExtensions(this.payload.signedExtensions);
 `
Need support for 

    `registry.setSignedExtensions(this.payload.signedExtensions, userExtensions);`

Also on our chain we override the definition of `ChargeTransactionPayment`, however during signing it uses the default type definition

**Comment by jacogr** [comment]
This is actually not solvable atm since the apps UI just doesn't know about these - and it is the injector for this information.

So to solve it requires a bit of work -

- the API typesBundle types needs to be adjusted to allow for signed extensions, specifically here https://github.com/polkadot-js/api/blob/master/packages/types/src/types/registry.ts#L47
- like it does with types, the API needs to expand the `userExtensions` alongside the types bundle and inject those for use, aka the same way as we do in the known types here for RPC https://github.com/polkadot-js/api/blob/master/packages/types-known/src/index.ts#L66

Once all that is done, the apps UI can be adjusted to inject the extensions as well, without that it will only always inject what is in the metadata, here https://github.com/polkadot-js/apps/blob/master/packages/page-settings/src/useChainInfo.ts#L26

After that, the extension can save this additionally and inject as required based on the chain.

This has not come up as a requirement, since the chains contributed in the apps UI above has not had that need (this is the provider of information). Sadly without that it is really working in a black box since there are no actual way of testing and getting it working end-to-end. 

TL;DR Certainly do-able, but not testable atm.

**Comment by KarishmaBothara** [comment]
Thanks a lot for your reply and the detailed explanation...It is very helpful...
Just one doubt I have now is... If I don't use the apps UI and have UI in which I am trying to use the extension with....i.e using web3Enable, web3AccountsSubscribe to talk to extension... Once API(polkadot) is upgraded with the changes you pointed out.. how can my UI pass that information to extension?

**Comment by jacogr** [comment]
The apps UI dictates the format at this point, so the extension doesn't know about it and would just ignore it completely. (As I mentioned, the apps UI is deemed as the primary provider of this information for extension upgrades and the 2 move in lock-step on the exchange formats)

I'm not against extending it at all, but at the same point it is not something that has been a requirement. (There are no known chains, to me, that it supported by the 2 that needs it as of yet. The fact that other UIs exchange info with the extension is certainly new info and not an original intent, since the exchange is not cast in stone as a "standard". There are quite  few issues remaining as it stands - great as a first-pass, but nowhere where it should be, there are actually some immediate breaking changes that I'm trying to get in for the next release, but breaking is also quite "tricky")

**Comment by holygits** [comment]
Hi @jacogr, I'm keen to pick this up. please let me know if this is on the right track / acceptable.

if I understand correctly we should update this definition to accept additional signed extensions and pass in from from the apps ui:
https://github.com/polkadot-js/extension/blob/a14651ab7f314f6a8995d5531be18e443e646989/packages/extension-inject/src/types.ts#L58

next here set userExtensions:
https://github.com/polkadot-js/extension/blob/a14651ab7f314f6a8995d5531be18e443e646989/packages/extension-chains/src/index.ts#L46

do we need to re-set signed extension types before every sign?
https://github.com/polkadot-js/extension/blob/253c9bf7858a5733134f8cc05ad847fd058013e8/packages/extension-base/src/background/RequestExtrinsicSign.ts#L19

Seems like we could add some decent tests within the extension itself on whether the custom signed extensions are picked up and used correctly or not.

**Comment by jacogr** [comment]
Ok, so on the API side, the `toCallsOnly` happens here, which is what the apps UI uses to pass it through (it is basically a stripped version of the metadata, i.e. stuff like constants/queries/events/errors are not needed and just add bloat) -

https://github.com/polkadot-js/api/blob/master/packages/metadata/src/util/toCallsOnly.ts (so it just needs to add the extrinsic info as well, the naming can stay

The above looks correct. 

Would be happy for PRs.
